Turn graphene on by default

ptschip opened the pull request:
This also required a couple of edge cases to be fixed in the request manager which were giving us trouble in the python tests.

Still needs work adding test cases to the request manager if possible. Also this introduces  a performance issue since now we have the check for if we have peers very frequently which lock cs_vNodes; I'll add an atomic flag in there so we don't have to cycle through vNodes so many times.

ptschip committed:
Turn graphene on by default

ptschip committed:
Check for HaveNodes

ptschip committed:
Handle edge case for requesting a full block.

Although very unlikely this edge case could happen if a node
decides not to run one of either thinblocks or graphene while at
the same time doesn't have either graphene peers or thinblock peers.

So for the following cases...

Graphene disable but Thinblock enabled, however there are graphene peers
but no thinblock peers connected.

or

Graphene enabled and Thinblocks disabled, however there are thinblock peers
but not graphene peers connected.

ptschip committed:
indent only - no semantic changes

gubatron commented on src/requestManager.cpp:
```diff
@@ -515,85 +515,90 @@ bool CRequestManager::RequestBlock(CNode *pfrom, CInv obj)
+    if (IsChainNearlySyncd() && IsThinBlocksEnabled() && HaveThinblockNodes())
     {
-        if (HaveThinblockNodes() && thindata.CheckThinblockTimer(obj.hash))
+        if (!IsGrapheneBlockEnabled() || (IsGrapheneBlockEnabled() && !graphenedata.CheckGrapheneBlockTimer(obj.hash)))
```
```cpp
if (!IsGrapheneBlockEnabled() || !graphenedata.CheckGrapheneBlockTimer(obj.hash))
```
redundant logic here, a single `IsGrapheneBlockEnabled()` check is sufficient.

sickpig replied:
@gubatron 

think you are right. good catch. :smile:

ptschip replied:
ok removed the redundant IsGrapheneBlockEnabled()

ptschip committed:
Prelim refactor of ipaddress() in bitcoin tests

Remove the repeating code for defining ipaddress() in the bitcoin
tests. Just have a global definition rather than having to create
multiple but slightly different definitions.

ptschip committed:
Add request manager tests for RequestBlock()

Make RequestBlock() public

Add helper functions to util.h: SetArg(), SetBoolArg()

Make GetTimeMillis() be useable if mocktime is set.  This is so we can test
the functionlity of the thinblock and graphene timers.

ptschip committed:
Remove redundant checked for IsGraphenBlockEnabled()

ptschip commented:
Taking out of WIP. Main issue was getting the request manager tests working to check all the code paths in RequestBlock()

ptschip committed:
fix format
